src/TheNextWeek/planar.h

```diff
@@ -0,0 +1,158 @@
+#ifndef PLANAR_H
+#define PLANAR_H
+//==============================================================================================
+// To the extent possible under law, the author(s) have dedicated all copyright and related and
+// neighboring rights to this software to the public domain worldwide. This software is
+// distributed without any warranty.
+//
+// You should have received a copy (see file COPYING.txt) of the CC0 Public Domain Dedication
+// along with this software. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//==============================================================================================
+
+#include "rtweekend.h"
+
+#include "hittable.h"
+
+#include <cmath>
+
+//// This class is somewhat complex, but that's because it serves as a base class for very
+//// simple derived classes, where you really only need to worry about construction, bounds,
+//// and UV hit determination.
+
+class planar : public hittable {
+  public:
+    planar(const point3& o, const vec3& side_A, const vec3& side_B, shared_ptr<material> m)
+      : origin(o), a(side_A), b(side_B), mat(m)
+    {
+        normal = unit_vector(cross(a, b));
+        D = -dot(normal, origin);
+        a /= dot(a,a);
+        b /= dot(b,b);
+
+/////// By default, bounded to the world coordinates of UV<0,0> to UV<1,1>.
+
+        interval u_bounds, v_bounds;
+        get_uv_bounds(u_bounds, v_bounds);
+        bbox = aabb(
+            origin + (u_bounds.min * side_A) + (v_bounds.min * side_B),
+            origin + (u_bounds.max * side_A) + (v_bounds.max * side_B)
+        );
+    }
+
+    virtual void get_uv_bounds(interval& u_bounds, interval& v_bounds) {
```

Comment:
what is this function doing?
It doesn't seem like anything overrides it?

Reply:
I was thinking of derived (possibly infinite) primitives, like an infinite strip, or something that tiled the infinite plane. Kind of related to my original idea that `planar` could itself be a primitive of the infinite plane. Of course, that brings up the issue of how you handle U,V coordinates (clamp{X,Y}, tile{X,Y}, mirror{X,Y}, ...).

Perhaps instead we should implement quad directly as a primitive, hardcoded to have [0,1] UV bounds, and then derived objects could have alternate UV hit methods but all be constrained to the UV in [0,1]×[0,1].

I was toying with the idea of replacing the big large sphere "floor" with an actual infinite plane, but just using a large quad would be sufficient. If I ever did my own version, I'd also toss in UV transforms.

Reply:
Since `planar` is a purely abstract class (since `hit_uv = 0`), I would rather not solve for the bbox at all.
I don't see the utility in it.

If the reader (or you) want to implement infinite strips or what have you, then it should be as a child class

Reply:
i.e.
```
 planar(const point3& o, const vec3& side_A, const vec3& side_B, shared_ptr<material> m)
      : origin(o), a(side_A), b(side_B), mat(m)
    {
        normal = unit_vector(cross(a, b));
        D = -dot(normal, origin);
        a /= dot(a,a);
        b /= dot(b,b);
}
```